09-intermediate-rails/api-creating-items.md

```diff
@@ -0,0 +1,220 @@
+</details>
+
+## Testing a Post Request
+
+In our API we can make a post request similar to our previous create actions
+
+```ruby
+  # pets_controller_test.rb
+  describe "create" do
+    let(:pet_data) {
+      {
+        pet: {
+          age: 13,
+          name: 'Stinker',
+          human: 'Grace'
+        }
+      }
+    }
+    it "can create a new pet" do
+      expect {
+        post pets_path, params: pet_data
+      }.must_differ 'Pet.count', 1
+
+      must_respond_with :created
+    end
+  end
+```
+
+As before, the big difference from a web app is in what we send back. While in a web app we would probably redirect the client to the resource they created, that doesn't make sense for an API. Instead, we just send back some JSON containing the created pet's id, as well as the appropriate status code.
+
+### Cross-Site Request Forgery (CSRF)
+
+**ALERT!** Rails is built with CSRF protection which essentially prevents malicious requests. You might remember seeing notes in passing about this in our forms curriculum, where we had to explicitly add an _authenticity token_ to each form.
+
+Since preventing these cross site requests doesn't apply in an API context, we will add a Rails helper method to our controller to allow these `POST` requests to go through:
+```ruby
+# pets_controller.rb
+class PetsController < ApplicationController
+  protect_from_forgery with: :null_session
+
+  ...
+```
+
+You can [read more about CSRF](http://guides.rubyonrails.org/security.html#cross-site-request-forgery-csrf) in the Rails documentation.
+
+### Creating A Resource
+
+We can now write our `create` method to make this test pass!
+
+```ruby
+  # pets_controller.rb
+  def create
+    pet = Pet.new(pet_params)
+
+    if pet.save
+      render json: pet.as_json(only: [:id), status: :created
+      return
+    else
+      render json: { ok: false, errors: pet.errors.messages }, status: :bad_request
+      return
+    end
+  end
+```
+
+### Sending POST Data Via Postman
+
+While we could use our browser to exercise the `index` and `show` endpoints, sending data to our API is a little trickier. For this, we'll use Postman.  Start up your rails server first!
+
+First, the setup. Select `POST` from the list of verbs and type in your endpoint's URI. Then, in the `Headers` tab, add a new key-value pair: `Content-Type` -> `application/json`. Postman should try to autocomplete these fields for you.
+
+![Setting up a Postman POST](images/10-creating-apis/Postman-POST-settings.png)
+
+Next, the data. Click the `Body` tab, select `raw`, and enter your JSON data in the text area.
+
+![Adding data to a Postman POST](images/10-creating-apis/Postman-POST-data.png)
+
+Click send, and see what the new pet ID is.
+
+### Handling Errors
+
+So far our pet creation endpoint assumes everything goes swimmingly, but this won't always be the case. For example, our Pet model has some validations - what happens if they fail? What should we send back to the client?
```

Comment:
It would be interesting if the lesson maintained the TDD format by not adding the "else" block to the source code until writing this test. But the content gets through just fine either way.

Reply:
Adjusted